### DISCUSSION TITLE
NO_PROXY is not working for IP range/mask

### DISCUSSION THREAD

**Discussion by harochau** [encode/httpx#2737] [Potential Issue]
looks like get_environment_proxies is working fine and sets no-proxy correctly, but it is ignored further down the line.
Interestingly, `aiohttp` shows the same faulty behavior and only `requests` making the correct requests
```python
import asyncio
import os

import aiohttp

import httpx
from httpx._utils import get_environment_proxies
import requests

os.environ.update({"HTTP_PROXY": "http://192.168.1.2:8000"})  # dummy proxy
os.environ.update({"NO_PROXY": "192.168.0.0/16"})
r = get_environment_proxies()
print(r)  # seems correct

httpx.get('http://192.168.1.2:8001')  # going to proxy
requests.get('http://192.168.1.2:8001')  # NOT going to proxy


async def amain():
    async with aiohttp.ClientSession(trust_env=True) as session:
        async with session.get(url='http://192.168.1.2:8001') as response:  # going to proxy
            print(await response.text())

asyncio.run(amain())

```

**Comment by karpetrosyan** [comment]
`HTTPX` does not support wildcards with ip ranges at all, but we should probably inform the client of this.

**Reply by HuM4NoiD** [reply]
Can such support be included? requests package supports ip range in the form `a.b.0.0/n`

**Reply by lovelydinosaur** [reply]
Points for consideration...

* Could you point to a good resource that neatly describes the IP netmask format?
* In order to have a cohesive API onto this we might consider this from the perspective of the `mounts={}` API, rather than the `NO_PROXY` environment variable.
* One way to add support here would be to implement this functionality using a custom transport, that performs the routing. Can you sketch out an implementation of this?

```python
subnet_mask = SubNetMask({
    "*": httpx.HTTPTransport(proxy="http://192.168.1.2:8000"),  # dummy
    "192.168.0.0/16": httpx.HTTPTransport())  # nah
})
client = httpx.Client(transport=subnet_mask)
```

Related:

* https://www.python-httpx.org/advanced/#mounting-transports
* https://github.com/encode/httpx/pull/2879/files
